sw/device/silicon_creator/rom_ext/rom_ext_epmp.c

```diff
@@ -6,173 +6,73 @@
-  //             +-----------+-----------+-----------+-----------+
-  // `pmpcfg1` = |  `pmp7cfg |  `pmp6cfg`| `pmp5cfg` | `pmp4cfg` |
-  //             +-----------+-----------+-----------+-----------+
-  CSR_WRITE(CSR_REG_PMPADDR7,
-            (uint32_t)region.start >> 2 |
-                ((uint32_t)region.end - (uint32_t)region.start - 1) >> 3);
-  CSR_CLEAR_BITS(CSR_REG_PMPCFG1, 0xff << 24);
-  CSR_SET_BITS(CSR_REG_PMPCFG1,
-               ((kEpmpModeNapot | kEpmpPermLockedNoAccess) << 24));
-}
-
-void rom_ext_epmp_clear_rom_region(void) {
-  const int kEntry = 2;
-  const uint32_t kMask = (0xFF << 16);
-  epmp_state.pmpaddr[kEntry] = 0;
-  epmp_state.pmpcfg[0] &= ~kMask;
-
-  // Update the hardware configuration (CSRs).
-  //
-  // Entry is hardcoded as 2. Make sure to modify hardcoded values if changing
-  // kEntry.
-  //
-  // The `pmp2cfg` configuration is the third field in `pmpcfg0`.
-  //
-  //            32          24          16           8           0
-  //             +-----------+-----------+-----------+-----------+
-  // `pmpcfg0` = |  `pmp3cfg |  `pmp2cfg`| `pmp1cfg` | `pmp0cfg` |
-  //             +-----------+-----------+-----------+-----------+
-  CSR_WRITE(CSR_REG_PMPADDR2, 0);
-  CSR_CLEAR_BITS(CSR_REG_PMPCFG0, kMask);
+void rom_ext_epmp_set_napot(uint8_t entry, epmp_region_t region,
+                            epmp_perm_t perm) {
+  uint32_t length = region.end - region.start;
+  // The length must be 4 or more.
+  HARDENED_CHECK_GE(length, 4);
+  // The length must be a power of 2.
+  HARDENED_CHECK_EQ(bitfield_popcount32(length), 1);
+  // The start address must be naturally aligned with length.
+  HARDENED_CHECK_EQ(region.start & (length - 1), 0);
+  epmp_mode_t mode = length == 4 ? kEpmpModeNa4 : kEpmpModeNapot;
+  uint32_t addr = (region.start >> 2) | ((length - 1) >> 3);
+  rom_ext_epmp_set(entry, (uint32_t)mode | (uint32_t)perm, addr);
 }
 
-void rom_ext_epmp_final_cleanup(void) {
-  // We want to clear the L bit from pmp3cfg and pmp4cfg to unlock the
-  // ROM_EXT region for the next stage.
-  //
-  //            32          24          16           8           0
-  //             +-----------+-----------+-----------+-----------+
-  // `pmpcfg0` = |  `pmp3cfg |  `pmp2cfg`| `pmp1cfg` | `pmp0cfg` |
-  //             +-----------+-----------+-----------+-----------+
-  // `pmpcfg1` = |  `pmp7cfg |  `pmp6cfg`| `pmp5cfg` | `pmp4cfg` |
-  //             +-----------+-----------+-----------+-----------+
-  const uint32_t kMask0 = (uint32_t)(EPMP_CFG_L << 24);
-  const uint32_t kMask1 = (uint32_t)(EPMP_CFG_L << 0);
-  epmp_state.pmpcfg[0] &= ~kMask0;
-  epmp_state.pmpcfg[1] &= ~kMask1;
-  CSR_CLEAR_BITS(CSR_REG_PMPCFG0, kMask0);
-  CSR_CLEAR_BITS(CSR_REG_PMPCFG1, kMask1);
+void rom_ext_epmp_set_tor(uint8_t entry, epmp_region_t region,
+                          epmp_perm_t perm) {
+  uint32_t start = region.start >> 2;
+  uint32_t end = ((region.end + 3u) & ~3u) >> 2;
+  rom_ext_epmp_set(entry, kEpmpModeOff, start);
```

Comment:
Something that may help to harden this to have a separate mechanism to update `epmp_state`. It becomes turtles all the way down, which is part of the reason why sec_mmio uses macros to update the counters. Okay to skip this comment, but maybe we should add a note saying that we are planning to harden this later.